recipes-bsp/u-boot/files/0011-board-siemens-iot2050-get-and-set-m.2-releated-infor.patch

```diff
@@ -0,0 +1,229 @@
+From ba758a41c49507b4abfcf292b695582f8cbb4375 Mon Sep 17 00:00:00 2001
+From: chao zeng <chao.zeng@siemens.com>
+Date: Fri, 7 Jan 2022 16:35:43 +0800
+Subject: [PATCH 11/16] board:siemens:iot2050:get and set m.2 releated
+ information
+
+Get the M.2 configuration,enable the power of the m.2
+Set the releated overlay according to the M.2 configuration
+
+Signed-off-by: chao zeng <chao.zeng@siemens.com>
+---
+ board/siemens/iot2050/board.c | 173 +++++++++++++++++++++++++++++++++-
+ 1 file changed, 171 insertions(+), 2 deletions(-)
+
+diff --git a/board/siemens/iot2050/board.c b/board/siemens/iot2050/board.c
+index 817fbf4ecb..21c82556e9 100644
+--- a/board/siemens/iot2050/board.c
++++ b/board/siemens/iot2050/board.c
+@@ -47,6 +47,141 @@ struct iot2050_info {
+ 
+ DECLARE_GLOBAL_DATA_PTR;
+ 
++#define M2_CONFIG_LEN    4
++#define PCIE_MUX_CONTROL 3
++#define SERDES_SELECT    2
++
++typedef struct gpio_config {
++	char *label;
++	char *gpio_name;
++	uint32_t direction;
++}gpioConfig_t;
++
++typedef enum lane0_config{
++	USB30 = 0,
++	PCIE0_LANE0,
++}lane0_config_e;
++
++typedef enum lane1_config{
++	PCIE1_LANE0 = 0,
++	PCIE0_LANE1,
++}lane1_config_e;
++
++typedef enum key_type{
++	M2_B_KEY = 0,
++	M2_BE_KEY,
++	M2_E_KEY,
++}key_type_e;
++
++typedef struct serdes_config {
++	lane0_config_e  lane0_sel;
++	lane1_config_e  lane1_sel;
++}serdes_config_t;
++
++typedef struct key_config {
++	key_type_e config_desc;    
++	int m2config[M2_CONFIG_LEN];
++	int m2_pinmux_value[PCIE_MUX_CONTROL];
++	serdes_config_t serdes_config;
++}key_config_t;
++
++
++static key_config_t config_table[] = {
++	{M2_B_KEY, {0,1,0,0}, {0,0,1}, {PCIE0_LANE0, PCIE0_LANE1}},
++	{M2_B_KEY, {0,0,0,1}, {0,1,0}, {PCIE0_LANE0, PCIE1_LANE0}},
++	{M2_B_KEY, {0,1,1,0}, {0,1,0}, {PCIE0_LANE0, PCIE1_LANE0}},
++	{M2_B_KEY, {1,0,0,1}, {0,1,0}, {PCIE0_LANE0, PCIE1_LANE0}},
++	{M2_B_KEY, {1,1,0,1}, {0,1,0}, {PCIE0_LANE0, PCIE1_LANE0}},
++	{M2_B_KEY, {0,0,0,1}, {1,1,0}, {USB30, PCIE1_LANE0}},
++	{M2_B_KEY, {0,1,0,1}, {1,1,0}, {USB30, PCIE1_LANE0}},
++	{M2_B_KEY, {0,0,1,1}, {1,1,0}, {USB30, PCIE1_LANE0}},
++	{M2_B_KEY, {1,0,1,1}, {1,1,0}, {USB30, PCIE1_LANE0}},
++};
++
++static gpioConfig_t m2_pinmux_config[PCIE_MUX_CONTROL] = {
++	{"SEL_USB30_PCIE0_LANE0", "gpio@600000_88", GPIOD_IS_OUT},
++	{"SEL_PCIE1_PCIE0",       "gpio@600000_82", GPIOD_IS_OUT},
++	{"SEL_USB30_PCIE0_LANE1", "gpio@600000_89", GPIOD_IS_OUT},
++};
++
++static gpioConfig_t m2_key_config[M2_CONFIG_LEN] = {
++	{"KEY_CONFIG_3", "gpio@601000_89", GPIOD_IS_IN},
++	{"KEY_CONFIG_2", "gpio@601000_88", GPIOD_IS_IN},
++	{"KEY_CONFIG_1", "gpio@601000_19", GPIOD_IS_IN},
++	{"KEY_CONFIG_0", "gpio@601000_18", GPIOD_IS_IN},
++};
```

Comment:
why not use the pinctrl in device tree?

Reply:
at lease the direction could be defined in the device tree.

Reply:
Plus above  gpio id comments, this definition could be reduced to some 

```
m2_b_key_config = {{0, 89}, {0, 88}, {1, 19}, {2, 18}}
```

Reply:
> why not use the pinctrl in device tree?

pinctrl in the device-tree set the pinmux and the direction.
we'd like to control the gpio under u-boot, we must request this resource.
and set or get the value.

Reply:
> at lease the direction could be defined in the device tree.

the direction could be deleted.